Track pull requests from king821221/datafu@master to linkedin/datafu@master

king821221 opened the pull request:
This pull request has centrally tracked the following features:

(1) create initial version of entroy UDFs
issue: https://github.com/linkedin/datafu/issues/77
(2) experiment and implement weighted sampling algorithms
issue: https://github.com/linkedin/datafu/issues/80

Commit:
create initial version of entroy UDFs

Commit:
update a few comments and error messages

Commit:
fix a bug in Entropy.accumulate() to use getFreq() method and update a few comments

matthayes commented:
Thanks for working on this!  I will start looking through it and reviewing the code.

matthayes commented on src/java/datafu/pig/stats/entropy/Entropy.java:
```diff
@@ -0,0 +1,350 @@
+ * <p>
+ * This UDF extends * {@link org.apache.pig.AccumulatorEvalFunc} and implements 
+ * * {@link org.apache.pig.Algebraic} so it could calculate entropy either in a streaming way 
+ * or in a distributed fashion levering combiner
```
typo: leveraging

king821221 replied:
Oh, thanks. Sorry I am not native English guy:)

matthayes commented on src/java/datafu/pig/stats/entropy/Entropy.java:
```diff
@@ -0,0 +1,350 @@
+package datafu.pig.stats.entropy;
```
needs header.  same goes for all files.  see other files as example.

matthayes replied:
This looks good now.

By the way, you could run ./check-license-headers.sh to verify all files have the header.

matthayes commented on src/java/datafu/pig/stats/entropy/Entropy.java:
```diff
@@ -0,0 +1,350 @@
+ * {@link <a href="http://en.wikipedia.org/wiki/Entropy_%28information_theory%29" target="_blank">wiki</a>} 
+ * 
+ * <p>
+ * This UDF extends * {@link org.apache.pig.AccumulatorEvalFunc} and implements 
```
some extra \* characters in this javadoc, please scrub

king821221 replied:
Thanks for your comment.

One question is how do I check the generated javadoc for this class?

matthayes replied:
Just run "ant" and the docs will be generated under docs/javadoc.

matthayes commented on src/java/datafu/pig/stats/entropy/Entropy.java:
```diff
@@ -0,0 +1,350 @@
+                
+                if(cxl == null || cxl.longValue() < 0) {
+                    //invalid input frequency
+                    t.set(0, null);
```
i wonder if we should throw an exception instead.  otherwise will people notice there is a problem?

king821221 replied:
This algebraic implementation follows AVG in pig built-in UDF: http://grepcode.com/file/repo1.maven.org/maven2/org.apache.pig/pig/0.11.1/org/apache/pig/builtin/AVG.java?av=f. It handles the invalid input as null.

matthayes replied:
I think it makes more sense to return an error in this case.  I'm not sure if ignoring invalid input is a good idea.

matthayes commented on src/java/datafu/pig/stats/entropy/Entropy.java:
```diff
@@ -0,0 +1,350 @@
+                }
+                return t;
+            } catch (NumberFormatException nfe) {
+                //invalid input format
```
again, not sure if we should ignore invalid input

matthayes commented on src/java/datafu/pig/stats/entropy/Entropy.java:
```diff
@@ -0,0 +1,350 @@
+            Double scxlogcx = (Double)t.get(0);
+            Long scx = (Long)t.get(1);
+            
+            sumOfCxLogCx += (scxlogcx == null ? 0 : scxlogcx.doubleValue());
```
would instead throw exception on null

matthayes commented on src/java/datafu/pig/stats/entropy/Entropy.java:
```diff
@@ -0,0 +1,350 @@
+        if(obj != null) {
+            switch (tp.getType(0))
+            {
+            case DataType.LONG : cx = (Long)obj; break;
```
why allow anything besides long and integer?

king821221 replied:
Hmm, I am not sure do we need such kind of logic to handle different input

Besides the use case I list in the javadoc, another possible case is that 1 team generates the term frequency and output to HDFS and another team reads the term frequency and generates entropy based on that. In this case, the entropy UDF indeed handles input maybe not explicit long or int, but bytearray or chararray.

matthayes replied:
To me only the int and long values make sense.  The user should convert to int or long themselves if they are not already of these types.

king821221 replied:
OK, will support only int and long.

matthayes commented on src/java/datafu/pig/stats/entropy/stream/ChaoShenEntropyEstimator.java:
```diff
@@ -0,0 +1,197 @@
+
+/*
+ * This class implements the Chao-Shen entropy estimator
+ * based on the estimator's implementation in R's entropy library
```
what is the license of the R implementation?

edit: 

looking at the documentation for the 'entropy' package, it is licensed as GPLv3:

http://cran.r-project.org/web/packages/entropy/entropy.pdf

GPLv3 is unfortunately not compatible with Apache 2.  If this code is a derivative of the R code then I don't think we can accept it.  We would have to do a clean implementation from scratch based on the paper:

http://www.researchgate.net/publication/226041346_Nonparametric_estimation_of_Shannons_index_of_diversity_when_there_are_unseen_species_in_sample/file/d912f510a79b8aae74.pdf

king821221 replied:
I see. Thanks for your notice. Will do as you suggest

matthayes commented on src/java/datafu/pig/stats/entropy/Entropy.java:
```diff
@@ -0,0 +1,350 @@
+                throw new RuntimeException("The field schema of the input tuple is null or its size is not 1");
+            }
+            
+            if(fieldSchemaList.get(0).type != DataType.BYTEARRAY &&
```
same comment on int and long

matthayes commented on src/java/datafu/pig/stats/entropy/stream/EmpiricalEntropyEstimator.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public void accumulate(long cx) {
+        N += cx;
```
simpler to wrap both lines in an if:

```
if (cx > 0) {
  ..
}
```

matthayes commented on src/java/datafu/pig/stats/entropy/stream/EntropyEstimator.java:
```diff
@@ -0,0 +1,89 @@
+           return new ChaoShenEntropyEstimator(base);
+        }
+        //unsupported estimator type
+        return null;
```
throw exception instead?

king821221 replied:
Hmm, yes, i think what you said makes sense. 
But what is your rule of thumb on error handling?

matthayes replied:
Simplicity is one thing I consider.  If the estimator is undefined an exception should ultimately be thrown somewhere.  Better to throw it here than check if the result is null each time you call this method and then throw an exception if so.

matthayes commented on src/java/datafu/pig/stats/entropy/Entropy.java:
```diff
@@ -0,0 +1,350 @@
+    @Override
+    public void cleanup()
+    {
+      if(this.streamEstimator != null) {
```
streamEstimator should never be null right?

king821221 replied:
Yes, so no need on this line

matthayes commented on src/java/datafu/pig/stats/entropy/stream/StreamingEntropy.java:
```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2010 LinkedIn, Inc
```
use: LinkedIn Corp. and contributors

king821221 replied:
OK, will do it

Is it for all files to follow this header?

matthayes replied:
Yep all java files, src and test.

matthayes commented on src/java/datafu/pig/stats/entropy/stream/EntropyEstimator.java:
```diff
@@ -0,0 +1,89 @@
+                
+                double base = Double.parseDouble(this.base);
+                
+                if(base > 0) {
```
what if base <=0 ?

king821221 replied:
yes, if base is non-positive, we fall back to Euler's number as the logarithm base.

The error checking is best to be put into the constructor? 

But what if user inputs "e", this should still be valid?

king821221 replied:
The base is now fixed to only accept "log", "log2" and "log10", which could avoid checking the base's numeric value. This is also what entropy library in R did

matthayes replied:
Okay this sounds good.

matthayes commented on src/java/datafu/pig/stats/entropy/stream/EntropyEstimator.java:
```diff
@@ -0,0 +1,89 @@
+                    h = h / Math.log(base);
+                }
+                
+            } catch (NumberFormatException ex) {
```
i would just throw an exception in this case, if they ask us for base "abc" we shouldn't proceed

matthayes commented on src/java/datafu/pig/stats/entropy/stream/StreamingEntropy.java:
```diff
@@ -0,0 +1,211 @@
+
+  public StreamingEntropy(String type, String base)
+  {
+    this.x = null;
```
call cleanup() instead, and move whatever code you need to there.  there is code in two places currently and that can be hard to maintain.

king821221 replied:
Yes, what you said makes sense. Will do it.

matthayes commented on src/java/datafu/pig/stats/entropy/stream/StreamingEntropy.java:
```diff
@@ -0,0 +1,211 @@
+  {
+    //do not miss the last tuple
+    try {
+        estimator.accumulate(cx);
```
i'm not sure how getValue() is supposed to be called, but I would make it work correctly even if it is called twice.  that is, set cx back to zero after you accumulate.

king821221 replied:
http://grepcode.com/file/repo1.maven.org/maven2/org.apache.pig/pig/0.11.1/org/apache/pig/AccumulatorEvalFunc.java?av=f

My understanding is that cleanup() will be called after getValue() for each invocation of exec().

matthayes replied:
Sure, but still better I think if getValue() was called more than once it would return the same value.

matthayes commented on test/pig/datafu/test/pig/stats/entropy/EntropyTests.java:
```diff
@@ -0,0 +1,548 @@
+  @Multiline private String entropy;
+  
+  @Test
+  public void uniqValEntropoyTest() throws Exception
```
typo in method name, fix others too

king821221 replied:
Oh, I will fix these typo errors

matthayes commented on test/pig/datafu/test/pig/stats/entropy/EntropyTests.java:
```diff
@@ -0,0 +1,548 @@
+    List<Double> expectedOutput = new ArrayList<Double>();
+    expectedOutput.add(2.302585);
+    
+    List<Tuple> output = this.getLinesForAlias(test, "data_out");
```
assert that output length matches expected length

also probably best to put this code into a reusable method since you use it in other places

king821221 replied:
OK

matthayes commented on test/pig/datafu/test/pig/stats/entropy/EntropyTests.java:
```diff
@@ -0,0 +1,548 @@
+  @Test
+  public void dupValEntropoyTest() throws Exception
+  {
+    System.out.println(entropy);
```
remove this as it will get a bit spammy, same for others as well

king821221 replied:
Do you mean removing "System.out.println(entropy)"?

matthayes replied:
Yea that's what I meant.  I'd like to avoid the console output getting too long because this can cause problems with travis.

matthayes commented on test/pig/datafu/test/pig/stats/entropy/EntropyTests.java:
```diff
@@ -0,0 +1,548 @@
+  @Multiline private String pairEntropy;
+ 
+  @Test
+  public void dupPairValEntropoyTest() throws Exception
```
this test seems redundant with dupValEntropoyTest

king821221 replied:
This is to test if the input tuple contains  more than 1 elements, it still works

matthayes replied:
But you are not really testing anything different when it comes to Entropy.  In both cases, just counts are passed in.  All you are testing is that GROUP followed by COUNT on tuples with more than 1 element works.

matthayes commented on test/pig/datafu/test/pig/stats/entropy/EntropyTests.java:
```diff
@@ -0,0 +1,548 @@
+  /**
+  register $JAR_PATH
+
+  define Entropy datafu.pig.stats.entropy.Entropy();
```
did you mean to pass in '2' ?

matthayes commented on test/pig/datafu/test/pig/stats/entropy/EntropyTests.java:
```diff
@@ -0,0 +1,548 @@
+  public void rawByteInputEntropoyTest() throws Exception
+  {
+    System.out.println(rawByteInputEntropy);
+    PigTest test = createPigTestFromString(rawByteInputEntropy); // logarithm base is 2 
```
or is this 2 a typo?

king821221 replied:
yes. will remove this comment. Thanks

matthayes commented on test/pig/datafu/test/pig/stats/entropy/EntropyTests.java:
```diff
@@ -0,0 +1,548 @@
+         test.runScript();
+         List<Tuple> output = this.getLinesForAlias(test, "data_out");
+    } catch(Exception ex) {
+         System.out.println(ex);
```
no need to print this i think.  also i think it's cleaner to fail at the end of the try block rather than at the end of the method, and change the exception you catch to be more specific.  could also check the message contains something you expect based on the expected error.

king821221 replied:
Indeed, the caught exception is "Fail to get alias for output iterator". The thrown exception is not caught in the test. Is it captured in the getLinesForAlias()?

matthayes replied:
No I don't think it is caught by getLinesForAlias().  What is the exception class name?

matthayes commented on test/pig/datafu/test/pig/stats/entropy/EntropyTests.java:
```diff
@@ -0,0 +1,548 @@
+    List<Tuple> output = this.getLinesForAlias(test, "data_out");
+    for (Tuple t : output)
+    {
+      assertTrue(t.isNull(0));
```
i don't think we should ignore invalid data

matthayes commented on test/pig/datafu/test/pig/stats/entropy/EntropyTests.java:
```diff
@@ -0,0 +1,548 @@
+  @Multiline private String accumulatedEntropy;
+
+  @Test
+  public void accumulatedEntropoyTest() throws Exception
```
i'm not sure if this actually guarantees that the accumulator version is called.

it would be good to implement at least one test of each for 1) accumulator, and 2) algebraic, where the methods are explicitly invoked instead of using the pig script test.  this ensures that both versions work.

king821221 replied:
yes, i have tried to print out some info messages in the accumulate() method and find that it is indeed called if the input data is ordered by. The other tests are tested with algebraic methodology.

matthayes replied:
Okay good to know.  I suppose this makes sense.  With the "order", the data would be sorted in the reducer, which implies algebraic can't be used.

Commit:
update entropy implementation following code review in https://github.com/linkedin/datafu/pull/78; add streaming conditional entropy; add samples to calculate mutual information using entropy UDF

Commit:
update javadocs

Commit:
update javadocs

king821221 commented:
The updates based on review include:
(1) update namings,  comments, coding styles, javadocs
(2) refactor the unit tests to make them more compact

Also, add a new UDF to calculate conditional entropy and give example on how to use it combined with streaming entropy to calculate mutual information.

Not included in the update:

(1) a clean-up of the Chao-Shen entropy

(2) The error handling logic on invalid input to the Entropy UDF

The original rationale is to permit a few errors in the input data so that the whole job will not fail since the exception thrown out will cause the job to restart 3 times before it is killed which might be expensive. Moreover, the entropy calculation does not need to be that very exact so permitting a few invalid data may still allow us to get a very close approximation of the true entropy.

Commit:
update the javadocs of streaming empirical entropy and streaming chaoshen entropy

Commit:
Merge branch 'master' of https://github.com/king821221/datafu

king821221 commented:
After review the paper, I do not change the present ChaoShen entropy estimator since it accords with the algorithm formula in the paper and it is implemented in a different way from the R's entropy library. What I did is to update the comments of the class to avoid confusion.

king821221 commented:
https://travis-ci.org/linkedin/datafu/builds/16217381 ?

Matt,

could you please help see what is the problem with this?

The error seems the unit test runs for a long time. 

Do you know where to get the running time per unit test?

king821221 commented:
The overall unit test seems grow a lot after I add more unit tests.

I will remove some unit tests to try to decrease the time.

matthayes commented:
Before you remove unit tests, let me try parallelizing the build as described here: http://about.travis-ci.org/docs/user/speeding-up-the-build/

I'm thinking that we can run a separate build per test directory.

matthayes commented:
Okay I've parallelized the build.  This has been merged into master.  There are now 11 separate builds, one for each folder.  Try pulling in the changes and allow travis to kick off another build.

matthayes commented on src/java/datafu/pig/stats/entropy/stream/StreamingCondEntropy.java:
```diff
@@ -0,0 +1,290 @@
+ * 
+ * @see StreamingEntropy
+ */
+@Nondeterministic
```
Since there is no randomness here I think this actually is deterministic, so can remove this tag Nondeterministic.  Same goes for all other instances.

king821221 replied:
Yes, will remove it

matthayes commented on src/java/datafu/pig/stats/entropy/stream/StreamingCondEntropy.java:
```diff
@@ -0,0 +1,290 @@
+ * </p> 
+ * 
+ * <p>
+ * This UDF's constructor definition and parameters are the same as that of * {@link datafu.pig.stats.entropy.stream.StreamingEntropy}
```
Remove the extra star in this line.  It'd be good to check the generated javadocs also to make sure they look good.

king821221 replied:
Yes, have generated the javadoc to check. 
Will remove the extra starts

matthayes commented on src/java/datafu/pig/stats/entropy/stream/StreamingCondEntropy.java:
```diff
@@ -0,0 +1,290 @@
+      //do not miss the last tuple
+      try {
+          this.combEstimator.accumulate(this.cxy);
+          this.condXEstimator.accumulate(this.cx);
```
set cxy and cx to zero after accumulating

king821221 replied:
The cleanup() function will be called immediately after getValue() in AccumulatorEvalFunc(), so we may not need to set them to 0 here

matthayes commented on src/java/datafu/pig/stats/entropy/stream/StreamingCondEntropy.java:
```diff
@@ -0,0 +1,290 @@
+    public void accumulate(Tuple input) throws IOException
+    {
+      for (Tuple t : (DataBag) input.get(0)) {
+
```
It'd be worth validating that the tuple does in fact have size 2 here

king821221 replied:
yes, it is done in the outputSchema() function.

```
        List<Schema.FieldSchema> fieldSchemaList = tupleSchema.getFields();

        if(fieldSchemaList == null || fieldSchemaList.size() != 2) {
            throw new RuntimeException("The field schema of the input tuple is null or its size is not 2");
        }
```

matthayes replied:
yea, true.  i was thinking it's worth as an added safety check, given that the tuple may not necessarily be consistent with the reported schema (e.g. as example of poorly written UDF producing input data).  maybe this is overkill though.

matthayes commented on src/java/datafu/pig/stats/entropy/stream/StreamingCondEntropy.java:
```diff
@@ -0,0 +1,290 @@
+                       " previous object type: " + DataType.findTypeName(prevObjType) + 
+                       ", present object type: " + DataType.findTypeName(presentObjType));
+               }
+               if(DataType.compare(prevObj,
```
I don't think the extra type checking you have above is really necessary.  Simply using DataType.compare(Object o1, Object o) should be sufficient.

king821221 replied:
The reason I then change to this is according to the javadoc of DataType.compare() as follows, the last sentence
"
public static int compare(Object o1,
                          Object o2,
                          byte dt1,
                          byte dt2)
Same as compare(Object, Object), but does not use reflection to determine the type of passed in objects, relying instead on the caller to provide the appropriate values, as determined by findType(Object).

Use this version in cases where multiple objects of the same type have to be repeatedly compared. 
"

matthayes replied:
Sure, but the implementation of compare(Object, Object) uses two findType calls, exactly as you are doing, so you are not avoiding reflection.

Commit:
update entropy udfs based on code review:
1. update conditional entropy's datatype.compare() method
2. entropy udf only accepts int and long

Commit:
Implement and experiment with different weighted sampling algorithms
https://github.com/linkedin/datafu/issues/80

matthayes commented:
Thanks for the weighted sampling implementation!  It looks pretty good to me.

Commit:
update weighted reservoir sampler constructor unit test

matthayes commented:
Okay this code looks pretty good to me.

DataFu has actually just recently been accepted into the Apache Incubator :)  Bad news is that we have a new git repo set up for it and now have to have patches submitted to the Apache DataFu JIRA instead.  Sorry for the inconvenience :(  I filed a JIRA already at https://issues.apache.org/jira/browse/DATAFU-2 , you just need to attach the patches to it.

This should be pretty easy to do though.  Here are the steps:

1) Update the license headers (as documented at http://www.apache.org/legal/src-headers.html ) like below.  Note there is no copyright anymore.

``` java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
```

2) Create the patches from your repo following the steps below:

``` bash
git remote add linkedin https://github.com/linkedin/datafu.git
git fetch linkedin
git pull linkedin master
git format-patch linkedin/master
```

This should ten or so patch files.  Just attach them to the JIRA I filed: https://issues.apache.org/jira/browse/DATAFU-2 .

Commit:
implement weighted reservoir sampling algorithm with exponential jumps

Commit:
update licence headers and remove un-used import classes

Commit:
update licence headers of entropy unit tests

matthayes commented:
Comments in the review board:

https://reviews.apache.org/r/17058/

matthayes commented:
Closing this as now being tracked in JIRA
